src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+    else
+        r = sol_buffer_append_printf(&buf, "http://%s", server);
+    SOL_INT_CHECK_GOTO(r, < 0, exit);
+
+    if (port != 80) {
+        r = sol_buffer_append_printf(&buf, ":%d", port);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    if (path && (strlen(path) >= 1 && path[0] != '/')) {
+        r = sol_buffer_append_printf(&buf, "%s%s",
+            path[0] != '/' ? "/" : "", path);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    if (encoded_params && *encoded_params) {
+        r = sol_buffer_append_printf(&buf, "?%s", encoded_params);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    if (fragment && *fragment) {
+        r = sol_buffer_append_printf(&buf, "#%s", fragment);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    *uri = sol_buffer_steal(&buf, NULL);
+exit:
+    sol_buffer_fini(&buf);
+    free(encoded_params);
+    return r;
+}
+
+SOL_API int
+sol_http_create_simple_uri(char **uri, const char *base_url,
+    const struct sol_http_param *params)
+{
+    char *encoded_params;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+    SOL_NULL_CHECK(base_url, -EINVAL);
+
+    if (params) {
+        r = sol_http_encode_params(&encoded_params, SOL_HTTP_PARAM_QUERY_PARAM,
+            params);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    if (asprintf(uri, "%s/%s%s", base_url, *encoded_params ? "?" : "",
+        encoded_params) < 0) {
+        r = -ENOMEM;
+        SOL_ERR("Could not create the URL with the encoded parameters."
+            "URL: %s, parameters: %s", base_url, encoded_params);
+    }
+    free(encoded_params);
+    return r;
+}
+
+static bool
+_is_sub_delimin(char c)
+{
+    if (c != '!' && c != '$' && c != '&' && c != '\''
+        && c != '(' && c != ')' && c != '*' && c != '+' && c != ','
+        && c != ';' && c != '=')
+        return false;
+    return true;
+}
+
+static bool
+_is_unreserved(char c)
+{
+    if (!isalnum(c) && c != '-' && c != '.' && c != '_' && c != '~')
+        return false;
+    return true;
+}
+
+static bool
+_is_pct_encoded(char c)
+{
+    if (c != '%')
+        return false;
+    return true;
+}
+
+static bool
+_is_pchar(char c)
+{
+    if (!_is_unreserved(c) && !_is_pct_encoded(c) &&
+        !_is_sub_delimin(c) && c != ':' && c != '@')
+        return false;
+    return true;
+}
+
+static int
+_get_protocol(const char *uri, struct sol_str_slice *protocol, char **end)
+{
+    struct sol_str_slice p;
+    size_t i;
+
+    p.data = uri;
+    p.len = 0;
+
+    for (i = 0; uri[i] && uri[i] != ':'; i++, p.len++) {
+        if (!isalnum(uri[i]) && uri[i] != '+' &&
+            uri[i] != '-' && uri[i] != '.') {
+            SOL_ERR("Invalid character at the URI's scheme: %c", uri[i]);
+            return -EINVAL;
+        }
+    }
+
+    if (uri[i] != ':') {
+        SOL_ERR("Could not find the \":\" scheme");
+        return -EINVAL;
+    }
+
+    SOL_DBG("URI Protocol: %.*s", SOL_STR_SLICE_PRINT(p));
+    *end = (char *)p.data + p.len + 1;
```

Comment:
this cast is bad. And you can force `protocol` to be given, do the `*end = p.data + p.len + 1` outside, when needed.